evm-node.md

```diff
@@ -0,0 +1,338 @@
+      requires ACCT in ACCTS orBool notBool #accountExists(ACCT)
+```
+
+-   `#getStorageData` loads the value for a single storage key of a specified account by its address and storage offset.
+    If the storage key has already been loaded or the account does not exist, it does nothing.
+
+```{.k .node}
+    syntax Int ::= #getStorageData ( Int , Int ) [function, hook(BLOCKCHAIN.getStorageData)]
+ // ----------------------------------------------------------------------------------------
+    rule <k> #lookupStorage ACCT INDEX => . ... </k>
+         <account>
+           <acctID>  ACCT                                                         </acctID>
+           <storage> STORAGE => STORAGE [ INDEX <- #getStorageData(ACCT, INDEX) ] </storage>
+           <origStorage> ORIGSTORAGE => ORIGSTORAGE [ INDEX <- #getStorageData(ACCT, INDEX) ] </origStorage>
+           ...
+         </account>
+      requires notBool INDEX in_keys(STORAGE)
+
+    rule <k> #lookupStorage ACCT INDEX => . ... </k>
+         <account>
+           <acctID> ACCT </acctID>
+           <storage> STORAGE:Map </storage>
+           ...
+         </account>
+      requires INDEX in_keys(STORAGE)
+
+    rule <k> #lookupStorage ACCT _ => . ... </k>
+      requires notBool #accountExists(ACCT)
+```
+
+-   `#getCode` loads the code for a specified account by its address. If the code has already been loaded, it does nothing.
+    If the account does not exist, it also does nothing.
+
+```{.k .node}
+    syntax String ::= #getCode ( Int ) [function, hook(BLOCKCHAIN.getCode)]
+ // -----------------------------------------------------------------------
+    rule <k> #lookupCode ACCT => . ... </k>
+         <account>
+           <acctID> ACCT </acctID>
+           <code> #unloaded(_) => #parseByteStackRaw(#getCode(ACCT)) </code>
+           ...
+         </account>
+
+    rule <k> #lookupCode ACCT => . ... </k>
+         <account>
+           <acctID> ACCT </acctID>
+           <code> _:ByteArray </code>
+           ...
+         </account>
+
+    rule <k> #lookupCode ACCT => . ... </k>
+      requires notBool #accountExists(ACCT)
+```
+
+-   `#getBlockhash(N)` returns the blockhash of the Nth most recent block, up to a maximum of 256 blocks.
+    It is used in the implementation of the BLOCKHASH instruction as seen below.
+
+```{.k .node}
+    syntax Int ::= #getBlockhash ( Int ) [function, hook(BLOCKCHAIN.getBlockhash)]
+ // ------------------------------------------------------------------------------
+    rule <k> BLOCKHASH N => #getBlockhash(N) ~> #push ... </k> <mode> NORMAL </mode> requires N >=Int 0 andBool N  <Int 256
+    rule <k> BLOCKHASH N => 0                ~> #push ... </k> <mode> NORMAL </mode> requires N  <Int 0  orBool N >=Int 256
+```
+
+```{.k .node}
+    rule <k> EXTCODEHASH ACCT => HASH ~> #push ... </k>
+         <account>
+           <acctID> ACCT </acctID>
+           <code> #unloaded(HASH) </code>
+           ...
+         </account>
```

Comment:
This can be done simpler by just adding a rule `keccak(#unloaded(HASH)) => HASH`, so that we don't need to override th eexisting rule for `EXTCODEHASH`, which has problems with synchronization.